document/src/main/java/com/yahoo/document/json/JsonReader.java

```diff
@@ -97,28 +101,52 @@ public ParsedDocumentOperation readSingleDocumentStreaming(DocumentOperationType
-            if ( ! FIELDS.equals(parser.getCurrentName()))
-                throw new IllegalArgumentException("expected field \"fields\", but got " + parser.getCurrentName());
+            Boolean create = null;
+            String condition = null;
+            ParsedDocumentOperation operation = null;
+            while (JsonToken.END_OBJECT != parser.nextValue()) {
+                switch (parser.getCurrentName()) {
+                    case FIELDS -> {
+                        documentParseInfo.fieldsBuffer = new LazyTokenBuffer(parser);
+                        VespaJsonDocumentReader vespaJsonDocumentReader = new VespaJsonDocumentReader(typeManager.getIgnoreUndefinedFields());
+                        operation = vespaJsonDocumentReader.createDocumentOperation(
+                                getDocumentTypeFromString(documentParseInfo.documentId.getDocType(), typeManager), documentParseInfo);
+
+                        if ( ! documentParseInfo.fieldsBuffer.isEmpty())
+                            throw new IllegalArgumentException("expected all content to be consumed by document parsing, but " +
+                                                               documentParseInfo.fieldsBuffer.nesting() + " levels remain");
+
+                    }
+                    case CONDITION -> {
+                        if ( ! JsonToken.VALUE_STRING.equals(parser.currentToken()) && ! JsonToken.VALUE_NULL.equals(parser.currentToken()))
+                            throw new IllegalArgumentException("expected string value for condition, got " + parser.currentToken());
+
+                        condition = parser.getValueAsString();
+                    }
+                    case CREATE_IF_NON_EXISTENT -> {
+                        create = parser.getBooleanValue(); // Throws if not boolean.
+                    }
+                    default -> {
+                        // We ignore stray fields, but need to ensure structural balance in doing do.
+                        if (parser.currentToken().isStructStart()) parser.skipChildren();
+                    }
+                }
+            }
 
-            if (JsonToken.START_OBJECT != parser.currentToken())
-                throw new IllegalArgumentException("expected start of \"fields\" object, got " + parser.currentToken());
+            if (null != parser.nextToken())
+                throw new IllegalArgumentException("expected end of input, got " + parser.currentToken());
 
-            documentParseInfo.fieldsBuffer = new LazyTokenBuffer(parser);
-            VespaJsonDocumentReader vespaJsonDocumentReader = new VespaJsonDocumentReader(typeManager.getIgnoreUndefinedFields());
-            ParsedDocumentOperation operation = vespaJsonDocumentReader.createDocumentOperation(
-                    getDocumentTypeFromString(documentParseInfo.documentId.getDocType(), typeManager), documentParseInfo);
+            assert null != operation: "VespaDocumentReader should throw on missing fields";
```

Comment:
Also, wouldn't it be possible to hit this case simply by omitting the `fields` object from the JSON payload? If so, it should probably be handled like an `IllegalArgumentException` rather than an invariant violation.

Reply:
Ah the point was that the VespaDocumentReader already does this check. There are unit tests for this. 
If this ever changes, the unit tests will also fail, because the exception is now an assertion error, instead of the excpted IAEx.

Reply:
I guess I could equally well have used a comment ... but that's a weaker statement.

Reply:
`VespaDocumentJsonReader` does check this, but from what I can see that particular code (`VespaDocumentJsonReader.readPut()` presumably, correct me if I'm wrong) will only be invoked _iff_ there is a `fields` object present in the first place. I.e. we have hit the `case FIELDS ->` above as part of parsing the JSON object, which is where we instantiate the reader.

I may simply be confused and not quite see where/how we have transitively ensured the presence of the `fields` object.

Reply:
Ah, no, you're right. It wasn't originally like that, but now it is.